### ISSUE TITLE
Question : Pre-commit script

### ISSUE THREAD

**Issue by genthalili** [mariodavid/cuba-component-data-import#81]
Is the Pre-commit Script executed after of before updating the associated value ? 
Let say the import is the associated with value `A`, but before committing I want to update to `B`, which means that the ID should be updated as well ?

**Comment by mariodavid** [comment]
It is executed right before the import happens:
https://github.com/mariodavid/cuba-component-data-import/blob/master/modules/core/src/de/diedavids/cuba/dataimport/service/GenericDataImporterServiceBean.groovy#L120

**Comment by genthalili** [comment]
But this is just before the import to the database, so if you modify the format value of the association in the pre-script, it's not taking in consideration : 
The entity creating is done at very beginning of the process : 
https://github.com/mariodavid/cuba-component-data-import/blob/master/modules/core/src/de/diedavids/cuba/dataimport/service/GenericDataImporterServiceBean.groovy#L51

**Comment by genthalili** [comment]
I realised that I need to use is similar to "Custom attribute binding script" but this is not really supporting the association field:
https://github.com/mariodavid/cuba-component-data-import/blob/master/modules/core/src/de/diedavids/cuba/dataimport/binding/AttributeBinderFactory.groovy

**Comment by mariodavid** [comment]
well, i think the custom binding script lets you have all the freedom that you need in order to change bind the given attribute in whatever way you want.

I don't really understand your case here. Do you want to make adjustments on the import data before the normal binding logic kicks in?
What is the downside of using the custom binding script then? Besides saving a little bit of code wo write? I mean this is more a one-time effort, right?

**Comment by genthalili** [comment]
Exactly, the need is to make adjustment on the import data before the normal bindings. Custom script fits perfectly except for associations.
The problem with associations is that I'dont know the UUID, so cannot create that "foreign" object on custom script..

**Comment by mariodavid** [comment]
Noting prevents you from using the `dataManager` within the custom binding script. Actually this was one of the main reason to create it in the first place.

It gets already passed in as a parameter into the script. 

You can also use the `SimpleDataLoader` bean if you want something more easy like [here](https://github.com/mariodavid/cuba-component-data-import/blob/05e95e36b2764d32f386eec1d4d70d7b32f43337/modules/core/src/de/diedavids/cuba/dataimport/service/UniqueEntityFinderServiceBean.groovy#L56) and just pass in your search criteria.

So should be just fine.

I think there are improvements to this app component (like proper logging) that a little bit more important then to improve the import process further, because with the custom scripts you have the freedom to do any complex logic you want.

**Comment by mariodavid** [comment]
There is a bug that was introduced through the UI changes we made in #74 that prevents the combination of "association" with "custom binding script" (see #92).

So this has to be fixed before 0.4.0.

But generally, here's an example of a custom binding script with SimpleDataLoader (in this case I choosed `MlbPlayer.team.code`):

```
import de.diedavids.cuba.dataimport.data.SimpleDataLoader
import de.diedavids.ddcdit.entity.mlb.MlbTeam
def simpleDataLoader = com.haulmont.cuba.core.global.AppBeans.get(SimpleDataLoader.NAME)

return simpleDataLoader.loadByProperty(MlbTeam, "name", "St. Louis Browns")
```
